### PR TITLE
adds Recurring.listRecurringDetails call

### PR DESCRIPTION
**Pull request by rinaldifonseca**

### CODE REVIEW

File: lib/adyen/rest/authorise_payment.rb [hunk 1]
```diff
@@ -185,6 +185,60 @@ def reauthorise_recurring_payment(attributes)
...
+
+          while !attributes["recurringDetailsResult.details.{%}.recurringDetailReference".gsub("{%}", index.to_s)].empty? do
+            response = { card: {} }
+            response[:recurring_detail_reference] = attributes["recurringDetailsResult.details.{%}.recurringDetailReference".gsub("{%}", index.to_s)]
+            response[:creation_date]              = attributes["recurringDetailsResult.details.{%}.creationDate".gsub("{%}", index.to_s)]
+            response[:variant]                    = attributes["recurringDetailsResult.details.{%}.variant".gsub("{%}", index.to_s)]
+            response[:card][:holder_name]         = attributes["recurringDetailsResult.details.{%}.card.holderName".gsub("{%}", index.to_s)]
+            response[:card][:expiry_month]        = attributes["recurringDetailsResult.details.{%}.card.expiryMonth".gsub("{%}", index.to_s)]
+            response[:card][:expiry_year]         = attributes["recurringDetailsResult.details.{%}.card.expiryYear".gsub("{%}", index.to_s)]
+            response[:card][:number]              = attributes["recurringDetailsResult.details.{%}.card.number".gsub("{%}", index.to_s)]
```
Comment (wvanbergen) [inline lib/adyen/rest/authorise_payment.rb]: There must be a nicer way of doing this. This is the first time an API that returns a list is implemented, so maybe some helper methods are missing to make this easier. Any suggestions?

File: test/functional/payment_authorisation_api_test.rb [hunk 2]
```diff
@@ -78,4 +78,15 @@ def test_recurring_payment_api_request
...
+
+  def test_list_recurring_details_api_request
+    response = @client.list_recurring_details(
+      recurring: { contract: "RECURRING" },
+      merchant_account: 'VanBergenORG',
+      shopper_reference: 'willem42',
+    )
+
+    assert response.references.any?
+    assert response.details.any?
```
Comment (wvanbergen) [inline test/functional/payment_authorisation_api_test.rb]: Can we make these assertions a bit more specific? What do you expect the items in `references` and `details` to look like?

**Comment by wvanbergen** [comment]
The CI failure can be ignored; it's just JRuby being shitty.

**Comment by wvanbergen** [comment]
Looks good; just a few questions. Thanks for contributing this!

File: lib/adyen/rest/authorise_payment.rb (hunk 1 above)
Comment (rinaldifonseca) [inline lib/adyen/rest/authorise_payment.rb]: Hi @wvanbergen,

There are two main things I dont like in this implementation. 1) be forced to use a loop in order to get all attributes 2) We dont have a hierarchy in the attributes.  

So, I came up with this new version where I extracted a method to be responsible for map a response list. Then, when we need to parse a new kind of response list we just need to map the attributes of the response we want, and pass to Adyen::REST::Response#map_response_list.

Please, take a look at this gist and tell me what do you think:  https://gist.github.com/rinaldifonseca/438933aeeb1d0ee2ced1

Thanks!

Comment (wvanbergen) [inline lib/adyen/rest/authorise_payment.rb]: A helper method like that would be great! Feel free to add it to the generic Response object to clean up this method here.

Comment (rinaldifonseca) [inline lib/adyen/rest/authorise_payment.rb]: Cool!  What generic Response object do you mean?

Comment (wvanbergen) [inline lib/adyen/rest/authorise_payment.rb]: the `Adyen::REST::Response` class.

Comment (rinaldifonseca) [inline lib/adyen/rest/authorise_payment.rb]: Yes. In my gist I defined the map_response_list method as protected in the Adyen::REST::Response. Can we move forward then? =)

Comment (wvanbergen) [inline lib/adyen/rest/authorise_payment.rb]: yup!

Comment (rinaldifonseca) [inline lib/adyen/rest/authorise_payment.rb]: Done =)

I also made the tests more specific.

**Comment by wvanbergen** [comment]
Awesome!

**Pull request merged**